python/cuml/thirdparty_adapters/adapters.py

```diff
@@ -0,0 +1,344 @@
+from cupy.sparse import csr_matrix as gpu_csr_matrix
+from cupy.sparse import csc_matrix as gpu_csc_matrix
+from cupy.sparse import csc_matrix as gpu_coo_matrix
+from scipy import sparse as cpu_sparse
+from cupy import sparse as gpu_sparse
+
+from numpy import ndarray as numpyArray
+from cupy import ndarray as cupyArray
+from cudf.core import Series as cuSeries
+from cudf.core import DataFrame as cuDataFrame
+from pandas import Series as pdSeries
+from pandas import DataFrame as pdDataFrame
+from numba.cuda import devicearray as numbaArray
+
+
+numeric_types = [
+    np.int8, np.int16, np.int32, np.int64,
+    np.uint8, np.uint16, np.uint32, np.uint64,
+    np.intp, np.uintp,
+    np.float32, np.float64,
+    np.complex64, np.complex128
+]
+
+
+def check_sparse(array, accept_sparse=False, accept_large_sparse=True):
+    """Checks that the sparse array is valid
+    Parameters
+    ----------
+    accept_sparse : string, boolean or list/tuple of strings (default=False)
+        String[s] representing allowed sparse matrix formats, such as 'csc',
+        'csr', etc. If the input is sparse but not in the allowed format,
+        it will be converted to the first listed format. True allows the input
+        to be any format. False means that a sparse matrix input will
+        raise an error.
+    accept_large_sparse : bool (default=True)
+        If a CSR, CSC, COO or BSR sparse matrix is supplied and accepted by
+        accept_sparse, accept_large_sparse=False will cause it to be accepted
+        only if its indices are stored with a 32-bit dtype.
+    Returns
+    -------
+    None or raise error
+    """
+    if accept_sparse is True:
+        return
+
+    err_msg = "This algorithm does not support the sparse " + \
+              "input in the current configuration."
+
+    is_sparse = cpu_sparse.issparse(array) or gpu_sparse.issparse(array)
+    if is_sparse:
+        if accept_sparse is False:
+            raise ValueError(err_msg)
+
+        if not accept_large_sparse:
+            if array.indices.dtype != cp.int32 or \
+               array.indptr.dtype != cp.int32:
+                raise ValueError(err_msg)
+
+        if isinstance(accept_sparse, (tuple, list)):
+            if array.format not in accept_sparse:
+                raise ValueError(err_msg)
+        elif array.format != accept_sparse:
+            raise ValueError(err_msg)
+
+
+def check_dtype(array, dtypes='numeric'):
+    """Checks that the input dtype is part of acceptable dtypes
+    Parameters
+    ----------
+    array : object
+        Input object to check / convert.
+    dtype : string, type, list of types or None (default="numeric")
+        Data type of result. If None, the dtype of the input is preserved.
+        If "numeric", dtype is preserved unless array.dtype is object.
+        If dtype is a list of types, conversion on the first type is only
+        performed if the dtype of the input is not in the list.
+    Returns
```

Comment:
Blank line needed before "Returns" or "Parameters" etc. This comes up in a lot of functions here.